### ISSUE TITLE
ExceptionUtil.rethrow wraps RuntimeException redundantly

### ISSUE THREAD

**Issue by mdogan** [hazelcast/hazelcast#17506]
With the change in https://github.com/hazelcast/hazelcast/pull/17212, `ExceptionUtil.rethrow()` started to wrap `RuntimeException` too. But this can lead to indeterminate number of wrapping, which makes it difficult to work with.

For example, following code creates a `HazelcastException` with a cause depth of `10`:

```java
Throwable t = new MemberLeftException();
for (int i = 0; i < 10; i++) {
    try {
        throw ExceptionUtil.rethrow(t);
    } catch (Throwable e) {
        t = e;
    }
}

while (t != null) {
    System.err.println(t.getClass().getName());
    t = t.getCause();
}
```

Output:
```
com.hazelcast.core.HazelcastException
com.hazelcast.core.HazelcastException
com.hazelcast.core.HazelcastException
com.hazelcast.core.HazelcastException
com.hazelcast.core.HazelcastException
com.hazelcast.core.HazelcastException
com.hazelcast.core.HazelcastException
com.hazelcast.core.HazelcastException
com.hazelcast.core.HazelcastException
com.hazelcast.core.HazelcastException
com.hazelcast.core.MemberLeftException
```

There are too many places in the code, we catch the `Throwable` and rethrow it via `ExceptionUtil.rethrow()`. Also in many places, we catch the exception and inspect for the type of the `Throwable` itself or its cause. But with this change it's unknown that how many layers of cause we should inspect to see the actual cause. 

For example, following `catch` block in `ClusterStateManager.changeClusterState(..)` fails to catch `MemberLeftException`, because it's wrapped by one more level inside `HazelcastException`:

```java
} catch (Throwable e) {
    tx.rollback();
    notifyAfterStateChange(oldState, requestedState, isTransient);
    if (e instanceof TargetNotMemberException || e.getCause() instanceof MemberLeftException) {
        throw new IllegalStateException("Cluster members changed during state change!", e);
    }
    throw ExceptionUtil.rethrow(e);
}
```

**Comment by sancar** [comment]
We did not see changing which layer the cause can be found as a problem. Now, it is clear that it was not a correct decision, because we are using the cause to decide the logic internally(our users also can use this info). And hiding the cause a couple of layers deep broke that logic in some places. 

So at this point, I have to revert the pr so that in our internals we will not mess with the cause hierarchy. In proxies, we can use AbstractInvocationFuture.joinInternal instead to have the caller stack trace info. There are a couple of downsides:
1. We will still mess up with the cause hierarchy for the hazelcast users. It will not be possible to write code that depends on the cause of an exception thrown by hazelcast proxies. 
2. There are lots of usages of rethrow/catch. It might not be possible to find and replace all places at once. But if we miss any, it will not be broken. We will be missing caller stack trace info in the same cases. 
3. We need to be aware that we should use `joinInternal` when call is from user thread. If a new code uses `rethrow/catch` again, it will again be lacking the caller stack trace info. 

Given that we are breaking the cause class hierarchy for the users, we might choose to
1. remove `tryWrapInSameClass` even from `joinInternal`. This means that we will not have any clue about the caller the stacktrace. 
OR
2. Put back exception stacktrace rewrite and solve the race problem by cloning the exception. We have removed this in 4.0 for the sake of making exceptions immutable. The pr that we made the change. https://github.com/hazelcast/hazelcast/pull/15441
Related issue. https://github.com/hazelcast/hazelcast/issues/6247
